# Guided Project: Profitable App Profiles for the App Store and Google Play Markets

### 1. Analyzing Mobile App Data

For this project, we'll pretend we're working as data analysts for a company that builds Android and iOS mobile apps. We make our apps available on Google Play and the App Store.

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means our revenue for any given app is mostly influenced by the number of users who use our app — the more users that see and engage with the ads, the better. Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

### 2. Opening and Exploring the Data

In the previous step, we outlined that our aim is to help our developers understand what type of apps are likely to attract more users on Google Play and the App Store. To do this, we'll need to collect and analyze data about mobile apps available on Google Play and the App Store.

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play

Collecting data for over 4 million apps requires a significant amount of time and money, so we'll try to analyze a sample of the data instead. To avoid spending resources on collecting new data ourselves, we should first try to see if we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our goals:

A data set containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018.
A data set containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017.
We'll start by opening and exploring these two data sets. To make them easier for you to explore, we created a function named explore_data() that you can repeatedly use to print rows in a readable way.

In [1]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
 

#### 2.1 Datasets opening

Opening of the two data sets we mentioned above, and save both as lists of lists.

In [2]:
from csv import reader

apple_apps_opened = open('AppleStore.csv',encoding="utf8")
android_apps_opened = open('googleplaystore.csv',encoding="utf8")


apple_apps_read = reader(apple_apps_opened)
android_apps_read = reader(android_apps_opened)

apple_apps = list(apple_apps_read)
android_apps = list(android_apps_read)




#### 2.2 Dataset exploration

Exploring both data sets using the explore_data() function

In [3]:
explore_data(apple_apps,0,5,rows_and_columns=True) # Exploring AppleStore.csv

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']


Number of rows: 7198
Number of columns: 17


In [4]:
explore_data(android_apps,0,5,rows_and_columns=True) # Exploring googleplaystore.csv

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10842
Number of columns: 13


#### 2.3 Columns that could help us with our analysis

In [5]:
print("googleplaystore : ",android_apps[0]) # Displaying the column names for googleplaystore.csv
print("\n")
print("AppleStore : ",apple_apps[0]) # Displaying the column names for googleplaystore.csv

googleplaystore :  ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


AppleStore :  ['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


##### 2.3.1 Identify the columns that could help us with our analysis for googleplaystore.csv

- The "Category" and "Type" columns are very usefull in order to detect what king of apps are the most used. 
- The "Rating" column show us the most used apps. 
- The "Price" colums is important too because we need to look for the free apps only

##### 2.3.2 Identify the columns that could help us with our analysis for AppleStore

- The "Prime_genre" column is very usefull in order to detect what king of apps are the most used. 
- The "user_rating" column show us the most used apps. 
- The "price" colums is important too because we need to look for the free apps only

### 3. Deleting Wrong Data

Recall that at our company, we only build apps that are free to download and install, and that are directed toward an English-speaking audience. This means that we'll need to:

- Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
- Remove apps that aren't free.

In [6]:
del(android_apps[10473])

### 4. Removing Duplicate Entries: Part One

If we explore the Google Play data set long enough or look at the discussions section, we'll notice some apps have duplicate entries. For instance, Instagram has four entries : 

In [7]:
for row in android_apps :
    name = row[0]
    
    if name == "Instagram" :
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Counting of the number of duplicates in googleplaystore :

In [8]:
duplicate_entries = []
unique_entries = []

for row in android_apps[1:] :
    name = row[0]
    
    if name in unique_entries :
        duplicate_entries.append(name)
        
    else :        
        unique_entries.append(name)
        
print("Number of duplicate entries : ",len(duplicate_entries))
print(duplicate_entries[:15])

Number of duplicate entries :  1181
['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.If we examine the rows we printed for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times. We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

### 5. Removing Duplicate Entries: Part Two

On the previous screen, we looped through the Google Play data set and found that there are 1,181 duplicates. After we remove the duplicates, we should be left with 9,659 rows:

In [9]:
print("Number of rows expected : ",len(android_apps[1:]) - len(duplicate_entries))

Number of rows expected :  9659


In [10]:
#Creation of a dictionary where each key is a unique app name and the corresponding dictionary value is 
#the highest number of reviews of that app.

reviews_max = {}


for element in android_apps[1:] :
    name = element[0]        
    n_reviews = float(element[3])
        
        
    if (name in reviews_max) and (reviews_max[name] < n_reviews) :
        reviews_max[name] = n_reviews
    elif name not in reviews_max :
        reviews_max[name] = n_reviews
        

In [11]:
len(reviews_max)

9659

In [12]:
reviews_max

{'Photo Editor & Candy Camera & Grid & ScrapBook': 159.0,
 'Coloring book moana': 974.0,
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps': 87510.0,
 'Sketch - Draw & Paint': 215644.0,
 'Pixel Draw - Number Art Coloring Book': 967.0,
 'Paper flowers instructions': 167.0,
 'Smoke Effect Photo Maker - Smoke Editor': 178.0,
 'Infinite Painter': 36815.0,
 'Garden Coloring Book': 13791.0,
 'Kids Paint Free - Drawing Fun': 121.0,
 'Text on Photo - Fonteee': 13880.0,
 'Name Art Photo Editor - Focus n Filters': 8788.0,
 'Tattoo Name On My Photo Editor': 44829.0,
 'Mandala Coloring Book': 4326.0,
 '3D Color Pixel by Number - Sandbox Art Coloring': 1518.0,
 'Learn To Draw Kawaii Characters': 55.0,
 'Photo Designer - Write your name with shapes': 3632.0,
 '350 Diy Room Decor Ideas': 27.0,
 'FlipaClip - Cartoon animation': 194216.0,
 'ibis Paint X': 224399.0,
 'Logo Maker - Small Business': 450.0,
 "Boys Photo Editor - Six Pack & Men's Suit": 654.0,
 'Superheroes Wallpapers | 4K Backgrounds': 

In [13]:
#Removing the duplicate rows due to reviews_max

android_clean = []
already_added = []

for element in android_apps[1:] :
    name = element[0]        
    n_reviews = float(element[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added) :
        android_clean.append(element)
        already_added.append(name)

    

In [14]:
len(android_clean)

9659

In [15]:
android_clean[0:15]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up'],
 ['Sketch - Draw & Paint',
  'ART_AND_DESIGN',
  '4.5',
  '215644',
  '25M',
  '50,000,000+',
  'Free',
  '0',
  'Teen',
  'Art & Design',
  'June 8, 2018',
  'Varies with device',
  '4.2 and up'],
 ['Pixel Draw - Number Art Coloring Book',
  'ART_AND_DESIGN',
  '4.3',
  '967',
  '2.8M',
  '100,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design;Creativity',
  'June 20, 2018',
  '1.1',
  '4.4 and up'],
 ['Paper flowers instructions',
  'ART_AND_DESIGN',
  '4.4',
  '167',
  '5.6M',
  '50,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'March 26, 2017

### 6. Removing Non-English Apps: Part One

In the previous step, we managed to remove the duplicate app entries in the Google Play data set. Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are directed toward an English-speaking audience. However, if we explore the data long enough, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience.

We're not interested in keeping these apps, so we'll remove them. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).

Behind the scenes, each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character 'a' is 97, character 'A' is 65, and character '爱' is 29,233. We can get the corresponding number of each character using the ord() built-in function.If an app name contains a character that is greater than 127, then it probably means that the app has a non-English name.



#### 6.1 Function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters

In [16]:
def string_filtering(string) :
    for car in string :
        if ord(car) > 127 :
            return False
    return True

#### 6.2 Checking whether these app names are detected as English or non-English for the strings below :

- 'Instagram'
- '爱奇艺PPS -《欢乐颂2》电视剧热播'
- 'Docs To Go™ Free Office Suite'
- 'Instachat 😜'

In [17]:
#Results

applications = ['Instagram',
'爱奇艺PPS -《欢乐颂2》电视剧热播',
'Docs To Go™ Free Office Suite',
'Instachat 😜']

for element in applications :
    print(string_filtering(element))

True
False
False
False


### 7. Removing Non-English Apps: Part Two

#### 7.1 Advanced function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters

In [18]:
def advanced_string_filtering(string) :
    i = 0
    for car in string :
        if ord(car) > 127 :
            i += 1
    if i > 3 :
        return False
    return True
    

#### 7.2 Use the new function to check whether some app names are detected as English or non-English

In [19]:
#Results

applications = ['Docs To Go™ Free Office Suite',
'Instachat 😜',
'爱奇艺PPS -《欢乐颂2》电视剧热播']

for element in applications :
    print(advanced_string_filtering(element))   

True
True
False


#### 7.3 Filter out non-English apps from both data sets. If an app name is identified as English, we append the whole row to a separate list.



In [20]:
#Filter out the non-English apps from android_apps

english_android_apps = []

for element in android_clean :
    name = element[0]
    if advanced_string_filtering(name) == True :
        english_android_apps.append(element)

In [21]:
#Filter out the non-English apps from apple_apps

english_apple_apps = []

for element in apple_apps[1:] :
    name = element[2]
    if advanced_string_filtering(name) == True :
        english_apple_apps.append(element)

#### 7.4 Exploring of the data sets and see how many rows we can compt for each dataset

In [22]:
print(len(english_android_apps))# Andoid dataset
print(len(english_apple_apps))# Apple dataset

9614
6183


### 8. Isolating the Free Apps

#### 8.1 Isolating the free apps in separate lists

In [23]:
google_free_apps = []
ios_free_apps = []

for element in english_android_apps :
    price = element[7]
    price = price.replace('$',"")
    if float(price) == 0 :
        google_free_apps.append(element)
        
for element in english_apple_apps :
    price = element[5]
    if float(price) == 0 :
        ios_free_apps.append(element)

#### 8.2 Checking the length of each datalist


In [24]:
print(len(google_free_apps))
print(len(ios_free_apps))

8864
3222


### 9. Most Common Apps by Genre: Part One

#### 9.0 Context into why we want to find an app profile that fits both the App Store and Google Play

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.

#### 9.1 Google Play inspection

Inspection of both data sets and identification of the columns we could use to generate frequency tables to find out what are the most common genres in each market :

In [25]:
explore_data(android_apps,0,1)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




We can use the columns 'Category' and 'Genres' to generate frequency tables to find out what are the most common genres in google play store.

#### 9.2 Apple Store inspection

In [26]:
explore_data(apple_apps,0,1)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




We can use the 'Prime_genre' column to generate frequency tables to find out what are the most common genres in the apple store.

### 10. Most Common Apps by Genre: Part Two

In the previous screen, we looked at our validation strategy for an app idea, and then we inspected the data sets to identify the columns that might be useful for finding out what the most common genres in each market are. Our conclusion was that we'll need to build a frequency table for the prime_genre column of the App Store data set, and for the Genres and Category columns of the Google Play data set.

We'll build two functions we can use to analyze the frequency tables:

One function to generate frequency tables that show percentages
Another function we can use to display the percentages in a descending order

#### 10.1.1 Create a function named freq_table() 

The function should return the frequency table (as a dictionary) for any column we want. The frequencies should also be expressed as percentages.

In [27]:
def freq_table(dataset,index) :
    
    frequences = {}
    i = 0
    
    for element in dataset :
        column = element[index]
        
        if column in frequences :
            frequences[column] += 1
            i += 1
            
        if column not in frequences :
            frequences[column] = 1
            i += 1
            
    for element in frequences :
        frequences[element] = (frequences[element]/i) * 100
            
    return frequences
        
        
        
    

#### 10.1.2 Frequency table of the columns prime_genre, Genres, and Category

In [28]:
freq_table(ios_free_apps,12) #Column prime_genre in ios_free_appsand

{'Productivity': 1.7380509000620732,
 'Weather': 0.8690254500310366,
 'Shopping': 2.60707635009311,
 'Reference': 0.5586592178770949,
 'Finance': 1.1173184357541899,
 'Music': 2.0484171322160147,
 'Utilities': 2.5139664804469275,
 'Travel': 1.2414649286157666,
 'Social Networking': 3.2898820608317814,
 'Sports': 2.1415270018621975,
 'Health & Fitness': 2.0173805090006205,
 'Games': 58.16263190564867,
 'Food & Drink': 0.8069522036002483,
 'News': 1.3345747982619491,
 'Book': 0.4345127250155183,
 'Photo & Video': 4.9658597144630665,
 'Entertainment': 7.883302296710118,
 'Business': 0.5276225946617008,
 'Lifestyle': 1.5828677839851024,
 'Education': 3.662321539416512,
 'Navigation': 0.186219739292365,
 'Medical': 0.186219739292365,
 'Catalogs': 0.12414649286157665}

In [29]:
freq_table(google_free_apps,1) #Column Category in google_free_apps

{'ART_AND_DESIGN': 0.6430505415162455,
 'AUTO_AND_VEHICLES': 0.9250902527075812,
 'BEAUTY': 0.5979241877256317,
 'BOOKS_AND_REFERENCE': 2.1435018050541514,
 'BUSINESS': 4.591606498194946,
 'COMICS': 0.6204873646209386,
 'COMMUNICATION': 3.2378158844765346,
 'DATING': 1.861462093862816,
 'EDUCATION': 1.1620036101083033,
 'ENTERTAINMENT': 0.9589350180505415,
 'EVENTS': 0.7107400722021661,
 'FINANCE': 3.7003610108303246,
 'FOOD_AND_DRINK': 1.2409747292418771,
 'HEALTH_AND_FITNESS': 3.0798736462093865,
 'HOUSE_AND_HOME': 0.8235559566787004,
 'LIBRARIES_AND_DEMO': 0.9363718411552346,
 'LIFESTYLE': 3.9034296028880866,
 'GAME': 9.724729241877256,
 'FAMILY': 18.907942238267147,
 'MEDICAL': 3.531137184115524,
 'SOCIAL': 2.6624548736462095,
 'SHOPPING': 2.2450361010830324,
 'PHOTOGRAPHY': 2.944494584837545,
 'SPORTS': 3.395758122743682,
 'TRAVEL_AND_LOCAL': 2.33528880866426,
 'TOOLS': 8.461191335740072,
 'PERSONALIZATION': 3.3167870036101084,
 'PRODUCTIVITY': 3.892148014440433,
 'PARENTING': 0.6

In [30]:
freq_table(google_free_apps,9) #Column Genres in google_free_apps

{'Art & Design': 0.5979241877256317,
 'Art & Design;Creativity': 0.06768953068592057,
 'Auto & Vehicles': 0.9250902527075812,
 'Beauty': 0.5979241877256317,
 'Books & Reference': 2.1435018050541514,
 'Business': 4.591606498194946,
 'Comics': 0.6092057761732852,
 'Comics;Creativity': 0.01128158844765343,
 'Communication': 3.2378158844765346,
 'Dating': 1.861462093862816,
 'Education': 5.347472924187725,
 'Education;Creativity': 0.04512635379061372,
 'Education;Education': 0.33844765342960287,
 'Education;Pretend Play': 0.056407942238267145,
 'Education;Brain Games': 0.033844765342960284,
 'Entertainment': 6.069494584837545,
 'Entertainment;Brain Games': 0.078971119133574,
 'Entertainment;Creativity': 0.033844765342960284,
 'Entertainment;Music & Video': 0.16922382671480143,
 'Events': 0.7107400722021661,
 'Finance': 3.7003610108303246,
 'Food & Drink': 1.2409747292418771,
 'Health & Fitness': 3.0798736462093865,
 'House & Home': 0.8235559566787004,
 'Libraries & Demo': 0.936371841155234

#### 10.1.2 Sorted frequency table of the columns prime_genre, Genres, and Category

In [31]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [32]:
display_table(ios_free_apps,12) #Prime genre

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


In [33]:
display_table(google_free_apps,1)#Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [34]:
display_table(google_free_apps,9)#Genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

### 11. Most Common Apps by Genre: Part Three

#### 11.1 Analyzing of the frequency table generated for the prime_genre column of the App Store data set

In [35]:
display_table(ios_free_apps,12) #Prime genre

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


##### What is the most common genre? What is the runner-up?

- Most common genre : 'Games' with almost 58% 
- Runner-up : 'Entertainment' avec environ 8%



##### Other patterns

We can see a 'lifestyle' pattern :

- Sports : 2.1415270018621975
- Music : 2.0484171322160147
- Health & Fitness : 2.0173805090006205
- Productivity : 1.7380509000620732
- Lifestyle : 1.5828677839851024

We can also see a 'professional' pattern : 

- Business : 0.5276225946617008
- Book : 0.4345127250155183
- Navigation : 0.186219739292365
- Medical : 0.186219739292365

#####  General impression : are most of the apps designed for practical purposes (education, shopping, utilities, productivity, lifestyle) or more for entertainment (games, photo and video, social networking, sports, music) ?

The most part of the apps are designed for entertainment.

##### Can we recommend an app profile for the App Store market based on this frequency table alone? If there's a large number of apps for a particular genre, does that also imply that apps of that genre generally have a large number of users?

We can't recommend an app aprofile for the App Store market based on an unique frequency table, we need more informations, moreover this frequency table is only based on specifics features from the original dataset(only english and free apps). The fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.



#### 11.2 Analyzing of the frequency table generated for the Category and Genres column of the Google Play data set

In [36]:
display_table(google_free_apps,1)#Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

##### What is the most common genre? What is the runner-up?(for category)

- Most common genre :Family
- Runner-up : Game

##### Comparing the patterns  for the Google Play market with those for the App Store market.


The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:



In [37]:
display_table(google_free_apps,9)#Genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

### 12. Most Popular Apps by Genre on the App Store

#### 12.1 Generation of a frequency table for the prime_genre column to get the unique app genres

In [64]:
freq = freq_table(ios_free_apps,12) #Column prime_genre

In [57]:
genre_and_avg_apps = []

for genre in freq :
    
    total = 0
    len_genre = 0
    
    for element in ios_free_apps :
        genre_app = element[12]
        
        if genre_app == genre :
            user_rating = element[6]
            total += float(user_rating)
            len_genre += 1

    avg_nb = total / len_genre
    genre_and_avg_apps.append((avg_nb,genre))
    
genre_and_avg_apps_sorted = sorted(genre_and_avg_apps,reverse=True)

In [58]:
genre_and_avg_apps_sorted

[(86090.33333333333, 'Navigation'),
 (74942.11111111111, 'Reference'),
 (71548.34905660378, 'Social Networking'),
 (57326.530303030304, 'Music'),
 (52279.892857142855, 'Weather'),
 (39758.5, 'Book'),
 (33333.92307692308, 'Food & Drink'),
 (31467.944444444445, 'Finance'),
 (28441.54375, 'Photo & Video'),
 (28243.8, 'Travel'),
 (26919.690476190477, 'Shopping'),
 (23298.015384615384, 'Health & Fitness'),
 (23008.898550724636, 'Sports'),
 (22788.6696905016, 'Games'),
 (21248.023255813954, 'News'),
 (21028.410714285714, 'Productivity'),
 (18684.456790123455, 'Utilities'),
 (16485.764705882353, 'Lifestyle'),
 (14029.830708661417, 'Entertainment'),
 (7491.117647058823, 'Business'),
 (7003.983050847458, 'Education'),
 (4004.0, 'Catalogs'),
 (612.0, 'Medical')]

##### 12.2 Analyzing the results , one app profile recommendation for the App Store

In [61]:
list_navigation = []

for element in ios_free_apps :
    
    genre = element[12]
    user_rating = element[6]
    
    if genre == 'Navigation' :
        list_navigation.append((int(user_rating),element[2]))
        
navigation_sorted = sorted(list_navigation,reverse = True)

In [62]:
navigation_sorted[0:3]

[(345046, 'Waze - GPS Navigation, Maps & Real-time Traffic'),
 (154911, 'Google Maps - Navigation & Transit'),
 (12811, 'Geocaching®')]

On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together.

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating. However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

### 13. Most Popular Apps by Genre on Google Play

#### 13.1 Generating a frequency table for the Category column of the Google Play data set to get the unique app genres

In [81]:
freq_category = freq_table(google_free_apps,1)

In [82]:
list_category_avg = []

for category in freq_category :
    total = 0
    len_category = 0
    
    for element in google_free_apps :
        category_app = element[1]
        
        if category_app == category :
            nb_install = element[5]
            nb_install = nb_install.replace('+','')
            nb_install = nb_install.replace(',','')
            nb_install = float(nb_install)
            total += nb_install
            len_category += 1
            
    avg_nb = total / len_category
    category_avg = (avg_nb,category)
    list_category_avg.append(category_avg)
    
list_category_avg_sorted = sorted(list_category_avg,reverse = True)

In [85]:
list_communication = []

for element in google_free_apps :
    
    genre = element[1]
    nb_install = element[5]
    nb_install = nb_install.replace('+','')
    nb_install = nb_install.replace(',','')
    nb_install = float(nb_install)
    
    if genre == 'COMMUNICATION' and nb_install < 100000000 :
        list_communication.append((float(nb_install),element[0]))
        
communication_sorted = sorted(list_communication,reverse = True)

In [86]:
communication_sorted

[(50000000.0, 'free video calls and chat'),
 (50000000.0, 'Zalo – Video Call'),
 (50000000.0, 'Mail.Ru - Email App'),
 (50000000.0, 'Dolphin Browser - Fast, Private & Adblock🐬'),
 (50000000.0, 'Contacts'),
 (50000000.0, 'CM Browser - Ad Blocker , Fast Download , Privacy'),
 (50000000.0, 'Azar'),
 (10000000.0, 'myMail – Email for Hotmail, Gmail and Outlook Mail'),
 (10000000.0, 'chomp SMS'),
 (10000000.0, 'ZenUI Dialer & Contacts'),
 (10000000.0, 'Xperia Link™'),
 (10000000.0, 'Whoscall - Caller ID & Block'),
 (10000000.0, 'WhatsCall Free Global Phone Call App & Cheap Calls'),
 (10000000.0, 'WhatsApp Business'),
 (10000000.0, 'WEB.DE Mail'),
 (10000000.0, 'Voxer Walkie Talkie Messenger'),
 (10000000.0, 'TouchPal Keyboard - Fun Emoji & Android Keyboard'),
 (10000000.0, 'Text SMS'),
 (10000000.0, 'Talkray - Free Calls & Texts'),
 (10000000.0, 'Talkatone: Free Texts, Calls & Phone Number'),
 (10000000.0, 'Puffin Web Browser'),
 (10000000.0, 'Psiphon Pro - The Internet Freedom VPN'),
 (1000